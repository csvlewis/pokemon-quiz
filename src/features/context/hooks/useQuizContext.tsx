import { useContext } from "react";
import { QuizContextType } from "@/features/context/types/QuizContextType";
import { QuizContext } from "@/features/context/components/QuizContext";

export function useQuizContext(): QuizContextType {
  return useContext<QuizContextType>(QuizContext);
}
