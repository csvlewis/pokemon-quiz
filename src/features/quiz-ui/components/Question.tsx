import { Box, Stack, Typography } from "@mui/material";
import { useQuizContext } from "@/features/context/hooks/useQuizContext";

export function Question(): React.ReactNode {
  const { currentQuestion, questions, quizInProgress } = useQuizContext();

  if (quizInProgress)
    return (
      <Stack alignItems="center" spacing={2}>
        <Typography>
          Question {currentQuestion}/{questions.length}
        </Typography>
        <Typography>{questions[currentQuestion - 1]?.text}</Typography>
        <Box
          component="img"
          sx={{
            maxHeight: { xs: 250 },
            maxWidth: { xs: 250 },
          }}
          src={questions[currentQuestion - 1]?.imageUrl}
        />
      </Stack>
    );
}
