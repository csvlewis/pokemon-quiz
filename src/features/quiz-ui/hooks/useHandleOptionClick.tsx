import { useQuizContext } from "@/features/context/hooks/useQuizContext";
import { Option } from "@/types/Option";

export function useHandleOptionClick(): (option: Option) => void {
  const { displayAnswer, selectedAnswer, setSelectedAnswer } = useQuizContext();

  return (option: Option) => {
    if (displayAnswer) return;
    if (option === selectedAnswer) {
      setSelectedAnswer({});
    } else {
      setSelectedAnswer(option);
    }
  };
}
