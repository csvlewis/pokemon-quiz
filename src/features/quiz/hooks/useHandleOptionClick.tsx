import { Option } from "@/types/Option";
import { useQuizContext } from "@/features/context/hooks/useQuizContext";

export function useHandleOptionClick(): (option: Option) => void {
  const { displayAnswer, selectedAnswer, setSelectedAnswer } = useQuizContext();

  return (option: Option) => {
    if (displayAnswer) return;
    if (option === selectedAnswer) {
      setSelectedAnswer({ text: "", isCorrect: false });
    } else {
      setSelectedAnswer(option);
    }
  };
}
