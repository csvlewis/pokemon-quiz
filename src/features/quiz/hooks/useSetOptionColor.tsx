import { OverridableStringUnion } from "@mui/types";

import { Option } from "@/types/Option";
import { useQuizContext } from "@/features/context/hooks/useQuizContext";

export function useSetOptionColor(): (
  option: Option
) => OverridableStringUnion<"error" | "primary" | "success"> {
  const { displayAnswer } = useQuizContext();

  return (option: Option) => {
    let color: OverridableStringUnion<"error" | "primary" | "success"> =
      "primary";

    if (displayAnswer) {
      if (!option.isCorrect) {
        color = "error";
      } else {
        color = "success";
      }
    }

    return color;
  };
}
