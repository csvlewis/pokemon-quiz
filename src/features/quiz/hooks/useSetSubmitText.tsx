import { useQuizContext } from "@/features/context/hooks/useQuizContext";

export function useSetSubmitText(): () => string {
  const { currentQuestion, displayAnswer, questions } = useQuizContext();

  return () => {
    let text: string = "Submit";

    if (displayAnswer) {
      if (currentQuestion >= questions.length) {
        text = "See Results";
      } else text = "Next Question";
    }

    return text;
  };
}
