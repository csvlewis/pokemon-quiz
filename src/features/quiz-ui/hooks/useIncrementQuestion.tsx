import { useQuizContext } from "@/features/context/components/QuizContextProvider";

export function useIncrementQuestion(): () => void {
  const {
    currentQuestion,
    questions,
    setCurrentQuestion,
    setDisplayAnswer,
    setSelectedAnswer,
    setShowResults,
    setQuizInProgress,
  } = useQuizContext();

  return () => {
    if (currentQuestion >= questions.length) {
      setQuizInProgress(false);
      setShowResults(true);
      return;
    }
    setCurrentQuestion(currentQuestion + 1);
    setDisplayAnswer(false);
    setSelectedAnswer({});
  };
}
