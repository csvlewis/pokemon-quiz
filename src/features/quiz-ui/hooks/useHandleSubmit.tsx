import { useQuizContext } from "@/features/context/hooks/useQuizContext";
import { useIncrementQuestion } from "@/features/quiz-ui/hooks/useIncrementQuestion";

export function useHandleSubmit(): () => void {
  const {
    correctAnswers,
    displayAnswer,
    selectedAnswer,
    setCorrectAnswers,
    setDisplayAnswer,
  } = useQuizContext();
  const incrementQuestion = useIncrementQuestion();

  return () => {
    if (displayAnswer) {
      incrementQuestion();
    } else {
      if (selectedAnswer.isCorrect) setCorrectAnswers(correctAnswers + 1);
      setDisplayAnswer(true);
    }
  };
}
