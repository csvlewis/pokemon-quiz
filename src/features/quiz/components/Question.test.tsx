import React from "react";
import { render } from "@testing-library/react";
import { useQuizContext } from "@/features/context/hooks/useQuizContext";
import { Question } from "./Question";

jest.mock("@/features/context/hooks/useQuizContext");

const mockedUseQuizContext = useQuizContext as jest.Mock;

describe("Question", () => {
  it("should render correctly", () => {
    mockedUseQuizContext.mockReturnValue({
      currentQuestion: 1,
      questions: [
        {
          text: "Question 1",
          imageUrl: "https://example.com/image1.png",
          options: [
            { text: "Option 1", isCorrect: false },
            { text: "Option 2", isCorrect: true },
            { text: "Option 3", isCorrect: false },
            { text: "Option 4", isCorrect: false },
          ],
        },
        {
          text: "Question 2",
          imageUrl: "https://example.com/image2.png",
          options: [
            { text: "Option 1", isCorrect: true },
            { text: "Option 2", isCorrect: false },
            { text: "Option 3", isCorrect: false },
            { text: "Option 4", isCorrect: false },
          ],
        },
      ],
      selectedAnswer: { text: "Option 1", isCorrect: false },
      quizInProgress: true,
    });

    const { container } = render(<Question />);
    expect(container).toMatchSnapshot();
  });
});
