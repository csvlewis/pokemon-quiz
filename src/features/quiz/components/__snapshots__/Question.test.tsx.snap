// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Question should render correctly 1`] = `
<div>
  <div
    class="MuiStack-root css-6rg9iv-MuiStack-root"
  >
    <p
      class="MuiTypography-root MuiTypography-body1 css-rizt0-MuiTypography-root"
    >
      Question 
      1
      /
      2
    </p>
    <p
      class="MuiTypography-root MuiTypography-body1 css-rizt0-MuiTypography-root"
    >
      Question 1
    </p>
    <img
      class="MuiBox-root css-1ri4hkh"
      src="https://example.com/image1.png"
    />
  </div>
</div>
`;
